core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +59,175 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
 	return FFA_OK;
 }
 
+static struct sp_shared_mem *get_sp_shared_mem_by_handle(struct sp_session *s,
+							 uint64_t handle)
+{
+	struct sp_shared_mem *ssm = NULL;
+
+	/*
+	 * FF-A Spec 8.10.2:
+	 * Each Handle identifies a single unique composite memory region
+	 * description that is, there is a 1:1 mapping between the two.
+	 * This means that there can only be one SP linked to a specific handle.
+	 */
+	SLIST_FOREACH(ssm, &s->mem_head, link) {
+		if (ssm->s_mem->mem_descr->global_handle == handle)
+			return ssm;
+	}
+	return NULL;
+}
+
+static TEE_Result add_mem_region_to_sp(struct ffa_mem_access *mem_acc,
+				       struct shared_mem *mem_share,
+				       uint64_t global_handle)
+{
+	struct ffa_mem_access_perm *access_perm =
+		&mem_acc->access_perm;
+	uint16_t endpoint_id = 0;
+	struct sp_session *s = NULL;
+
+	endpoint_id = READ_ONCE(access_perm->endpoint_id);
+	s = sp_get_session(endpoint_id);
+
+	/* Only add memory shares of loaded SPs */
+	if (s) {
+		struct sp_shared_mem *ssm = NULL;
+
+		ssm = calloc(1, sizeof(struct sp_shared_mem));
+
+		if (!ssm)
+			return FFA_NO_MEMORY;
+
+		ssm->access_descr = mem_acc;
+
+		ssm->counter = 0;
+		ssm->endpoint_id = s->endpoint_id;
+
+		/* Only allow each endpoint ones */
+		if (get_sp_shared_mem_by_handle(s, global_handle)) {
+			free(ssm);
+			return FFA_DENIED;
+		}
+		/*
+		 * Make a link between struct shared_mem and
+		 * struct sp_shared_mem
+		 */
+		ssm->s_mem = mem_share;
+		SLIST_INSERT_HEAD(&s->mem_head, ssm, link);
+		SLIST_INSERT_HEAD(&mem_share->sp_head, ssm,
+				  link);
+
+	} else {
+		/*
+		 * We don't register memory that is shared with
+		 * the OP-TEE endpoint. However we do need to
+		 * check that if someone tries to share with an
+		 * invalid endpoint.
+		 */
+		if (endpoint_id != spmc_get_id())
+			return FFA_DENIED;
+	}
+
+	return FFA_OK;
+}
+
+TEE_Result spmc_sp_add_share(struct ffa_mem_transaction *input_descr,
+			     uint64_t global_handle, size_t blen)
+{
+	int rc = FFA_INVALID_PARAMETERS;
+	struct ts_session *ts = 0;
+	uint16_t caller_id = FFA_NW_ID;
+	unsigned int num_mem_accs = 0;
+	unsigned int i = 0;
+	void *mem_share_buffer = NULL;
+	struct ffa_mem_access *mem_acc = NULL;
+	struct shared_mem *mem_share = NULL;
+
+	ts = sp_get_active();
+
+	/* Find the endpoint which is sharing the memory region */
+	if (ts) {
+		struct sp_ctx *uctx = NULL;
+		struct sp_session *calling_s = NULL;
+
+		uctx = to_sp_ctx(ts->ctx);
+		calling_s = uctx->open_session;
+		caller_id = calling_s->endpoint_id;
+	}
+
+	num_mem_accs = READ_ONCE(input_descr->mem_access_count);
+
+	if (num_mem_accs) {
+		/*
+		 * We store the full incoming transaction buffer and have
```

Comment:
This seems to be wasting a bit of memory and also perhaps a greater risk of accessing unchecked fields.